### PR TITLE
Various commonly needed utility functions

### LINKED ISSUE #1: Utility functions for commonly occurring problems
First of all, I would like to thank you for your work implementing dkm. I have used it in one of my projects and it was a breeze to set up and use.

During the time I was in the process of incorporating dkm into the project, I have written several utility functions that helped me use the library more efficiently. Currently, I have functions to
- calculate distance of each point to cluster center
- calculate sum of distances of each point to cluster center
- return all points that belong to the given cluster
- calculate the inertia of a given clustering using sum of distances method
- return the best clustering based on inertia
- calculate kmeans by running dkm::kmeans_lloyd function a given number of times and return the best clustering

Would you like me to send a pull request with these functions added to **dkm.hpp** file under a **util** namespace?

P.S.: All the functions are template functions; so the library would still be header-only.

### PR DESCRIPTION
**Pull request by eozd**
This PR is about implementing commonly needed utility functions.

Fixes #1 .

## Proposed Changes
1. A separate ``distance`` function under ``dkm::details``
2. Following utility functions under namespace ``dkm``
  * **dist_to_center:** Distance of each point to its cluster center
  * **sum_dist:** Sum of distances of each point to its cluster center
  * **get_cluster:** Get points that belong to the cluster with the given label
  * **means_inertia:** Calculate Euclidean distance inertia of a given clustering
  * **get_best_means:** Return best means from a means list based on inertia
  * **n_kmeans:** Run k-means algorithm a given number of times and return the best clustering.

### CODE REVIEW

**Comment by genbattle** [comment]
I'll have another look tomorrow when I get some more time after work :-)

File: include/dkm_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ *
+ * @param points    Poinst to be sent to dkm::kmeans_lloyd.
+ * @param k			Number of clusters.
+ * @param n_init    Number of times dkm::kmeans_lloyd algorithm will be run.
+ *
+ * @return clustering with the lowest inertia.
+ */
+template <typename T, size_t N>
+std::tuple<std::vector<std::array<T, N>, std::vector<uint32_t>>>
+n_kmeans(const std::vector<std::array<T, N>>& points, uint32_t k, uint32_t n_init=10)
```
Comment (genbattle) [inline include/dkm_utils.hpp]: See my comment on `get_best_means` above. If you change `get_best_means` as suggested, this function will no longer be required.

**Comment by eozd** [comment]
I have made the changes as you requested. One thing I was unsure of is what you've meant about ``distance`` function. Do you mean rewriting ``distance_squared`` completely so that now it returns the actual distance instead of its square? If not, can you clarify so that I change it?

File: include/dkm_utils.hpp [hunk 2]
```diff
@@ -9,6 +9,41 @@
...
 namespace dkm {
 
+
+/**
+ * Calculates the Euclidean distance from each point in the given sequence
+ * to given center and writes the results onto the sequence starting
+ * with out.
+ *
+ * @param points Point sequence.
+ * @param out    Beginning of the output distance sequence.
```
Comment (genbattle) [inline include/dkm_utils.hpp]: These comments are now incorrect after the changes to the function parameter list.

File: include/dkm_utils.hpp [hunk 3]
```diff
@@ -44,85 +75,53 @@ std::vector<std::array<T, N>> get_cluster(const std::vector<std::array<T, N>>& p
  *
  * @param points Sequence that were passed to dkm::kmeans_lloyd
  * @param points Result of dkm::kmeans_lloyd
```
Comment (genbattle) [inline include/dkm_utils.hpp]: I think the parameter name in this comment is incorrect? Should this be `means`?

File: src/test/test.cpp [hunk 4]
```diff
@@ -271,68 +266,11 @@ const lest::test specification[] = {
...
-				{57.67046396,  35.05019015}
-			};
-			SECTION("Invoke dkm k-means with 5 clusters 10 separate times") {
-				std::vector<std::tuple<std::vector<std::array<double, 2>>, std::vector<uint32_t>>> means_list;
-				for (size_t i = 0; i < 10; ++i) {
-					means_list.push_back(dkm::kmeans_lloyd(points, 5));
-				}
-
-				// best means via dkm_utils
-				auto best_means = dkm::get_best_means(points, means_list);
```
Comment (genbattle) [inline src/test/test.cpp]: Why delete this test entirely instead of modifying it to work with the new get_best_means function?

**Pull request merged**